### ISSUE TITLE
CoreML export with a large number of classes produces very large model file and very slow inference

### ISSUE THREAD

**Issue by ALMerrill** [ultralytics/ultralytics#3331]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am working on exporting a YOLOv8 model to CoreML. It is a detection model, and it has about 3200 classes. Both the pytorch model file and the exported CoreML model file are both very large (I'll include a table below comparing all file sizes), and when I load the model into Xcode and run a performance test it fails, I'm guessing due to a timeout or something. When I load the model and run it on an iPhone or iPad, the model loading and inference time are at least 10x slower than a model with around 300 classes, which was never the case with YOLOv5. The inference time increased marginally with a higher number of classes. I also tried removing the optimizer as well as `ema` from the checkpoint object using pytorch, which did decrease the pytorch file size, but had no impact on the converted CoreML model.

| # of Classes  | Pytorch file | Pytorch file with optimizer and EMA stripped | CoreML file (regardless of pytorch changes) | 
| ------ | -------- | ------- | ---------- |
| 300    | 131 MB  |  273 MB |  261 MB  |
| 3200  | 3.5 GB  |  875MB  |  1.75 GB  |

I'm just trying to figure out what changed in YOLOv8 that would so drastically increase the file size in both pytorch and CoreML, and increase the inference time in CoreML. Is there anything I can do before I export the model that can remedy these issues?

Here's the code I've been using to test this
```
from ultralytics import YOLO
import torch

model = YOLO("last.pt")
# Try converting unchanged model
model.export(format="coreml", nms=True)
# I didn't write this down, but the export for the 3200 class model here took well over an hour

model.ckpt["ema"] = None
model.ckpt["optimizer"] = None
torch.save(model.ckpt, "stripped.pt")

# Try converting stripped model
model  = YOLO("stripped.pt")
model.export(format="coreml", nms=True)
# This export took 7.2 minutes with the 3200 class model
```

One other thing. I got this error at the end of the export for the 3200 class model that I did not strip the optimizer nor EMA from:
```
RuntimeWarning: You will not be able to run predict() on this Core ML model. Underlying exception message was: Error compiling model: "compiler error:  Encountered an error while compiling a neural network model: I/O error".
```

However, I was still able to perform inference, it was just very slow

**Comment by ALMerrill** [comment]
Quick update. I've been playing around with this some more. From what I've heard and tried, it seems like something in the model itself, maybe the convolutional layers, is affected significantly by the number of classes, causing these issues to happen when you reach a large class set size. I'd be curious to hear if there's any thoughts on how important those changes are, and if there'd be a way to make small changes to avoid that modeling decision, in order to benefit from the v8 and anchorless improvements, without the model size/memory/inference time getting out of control.

**Comment by glenn-jocher** [comment]
Yes this is caused by code in the head itself, all other parts of model are not affected by class count. 

The question is if we can modify the architecture while maintaining backwards compat with existing models.

**Comment by glenn-jocher** [comment]
BTW the observed results are not the intention, the intention is to scale params similarly as v5 does with classes, so we need to fix this.

**Comment by glenn-jocher** [comment]
@ALMerrill great news 😃! Your original issue may now be resolved ✅ in PR #3436. This PR caps the head parameter scaling for the classes split to a maximum of 100 (just slightly higher than COCO, so very large class-count datasets should now scale similarly to YOLOv5. This fix should be live in `ultralytics` 8.0.125 in about 24 hours.

You will need to train a new model to see this change, and I believe this should be backwards compatible with existing models.

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by ALMerrill** [comment]
That's great, thank you! I'll definitely try that out when it's available and report back.

**Comment by glenn-jocher** [comment]
@ALMerrill thank you for your feedback! We appreciate your patience. We have made the necessary changes in the YOLOv8 model to address the issues related to the number of classes causing problems. The update will be available in the next release version (v8.0.125) of YOLOv8. We recommend training a new model using the updated version to see the improvements. We look forward to hearing about your experience with the new version. Let us know if you have any further questions or concerns. Happy training with YOLOv8!
